server/app/auth/CiviFormProfileMerger.java

```diff
@@ -0,0 +1,69 @@
+package auth;
+
+import java.util.Optional;
+import java.util.function.BiFunction;
+import javax.inject.Provider;
+import models.Account;
+import models.Applicant;
+import org.pac4j.core.profile.UserProfile;
+import repository.UserRepository;
+
+/** This class serves to abstract out common user profile merging logic into a shared helper. */
+public class CiviFormProfileMerger {
+
+  private final ProfileFactory profileFactory;
+  private final Provider<UserRepository> applicantRepositoryProvider;
+
+  public CiviFormProfileMerger(
+      ProfileFactory profileFactory, Provider<UserRepository> applicantRepositoryProvider) {
+    this.profileFactory = profileFactory;
+    this.applicantRepositoryProvider = applicantRepositoryProvider;
+  }
+
+  /**
+   * This method handles a three way merge between an existing applicant, an existing guest profile,
+   * and an external profile
+   *
+   * @param applicantInDatabase a potentially existing applicant in the database
+   * @param sessionProfile a guest profile from the browser session
+   * @param authProviderProfile profile data from an external auth provider, such as OIDC
+   * @param mergeFunction a function that merges an external profile into a Civiform profile, or
+   *     provides one if it doesn't exist
+   */
+  public <T> Optional<UserProfile> threeWayMerge(
```

Comment:
I would be nice to have clearer names here.  "threeWay" doesn't add much beyond looking at the parameters.

Just merge() is good.

Also if you can provide any extra method docs to help the reader understand the interaction among the 4 parameters, that'd be great!